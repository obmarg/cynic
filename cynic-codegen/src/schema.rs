use crate::{FieldArgument, TypeIndex};

// Alias all the graphql_parser schema types so we don't have to specify generic parameters
// everywhere
pub type Document = graphql_parser::schema::Document<'static, String>;
pub type Type = graphql_parser::schema::Type<'static, String>;
pub type Field = graphql_parser::schema::Field<'static, String>;
pub type Definition = graphql_parser::schema::Definition<'static, String>;
pub type TypeDefinition = graphql_parser::schema::TypeDefinition<'static, String>;
pub type ObjectType = graphql_parser::schema::ObjectType<'static, String>;
pub type EnumType = graphql_parser::schema::EnumType<'static, String>;
pub type InterfaceType = graphql_parser::schema::InterfaceType<'static, String>;
pub type InputObjectType = graphql_parser::schema::InputObjectType<'static, String>;
pub type ScalarType = graphql_parser::schema::ScalarType<'static, String>;
pub type UnionType = graphql_parser::schema::UnionType<'static, String>;
pub type InputValue = graphql_parser::schema::InputValue<'static, String>;
pub type EnumValue = graphql_parser::schema::EnumValue<'static, String>;

/// Loads a schema from a filename, relative to CARGO_MANIFEST_DIR if it's set.
pub fn load_schema(filename: impl AsRef<std::path::Path>) -> Result<Document, SchemaLoadError> {
    use std::path::PathBuf;
    let mut pathbuf = PathBuf::new();

    if let Ok(manifest_dir) = std::env::var("CARGO_MANIFEST_DIR") {
        pathbuf.push(manifest_dir);
    } else {
        pathbuf.push(std::env::current_dir()?);
    }
    pathbuf.push(filename);

    let schema = std::fs::read_to_string(&pathbuf)
        .map_err(|_| SchemaLoadError::FileNotFound(pathbuf.to_str().unwrap().to_string()))?;

    parse_schema(&schema)
}

pub(crate) fn parse_schema(schema: &str) -> Result<Document, SchemaLoadError> {
    let borrowed_schema = graphql_parser::schema::parse_schema::<String>(&schema)?;
    Ok(schema_into_static(borrowed_schema))
}

fn schema_into_static<'a>(
    doc: graphql_parser::schema::Document<'a, String>,
) -> graphql_parser::schema::Document<'static, String> {
    // A workaround until https://github.com/graphql-rust/graphql-parser/pull/33 is
    // merged upstream.

    // A document that uses Strings for it's data should be safe to cast to 'static lifetime
    // as there's nothing inside it to reference 'a anyway
    unsafe { std::mem::transmute::<_, Document>(doc) }
}

#[derive(Debug, PartialEq, Clone)]
pub enum SchemaLoadError {
    IoError(String),
    ParseError(String),
    FileNotFound(String),
}

impl SchemaLoadError {
    pub fn into_syn_error(self, schema_span: proc_macro2::Span) -> syn::Error {
        let message = match self {
            SchemaLoadError::IoError(e) => format!("Could not load schema file: {}", e),
            SchemaLoadError::ParseError(e) => format!("Could not parse schema file: {}", e),
            SchemaLoadError::FileNotFound(e) => format!("Could not find file: {}", e),
        };

        syn::Error::new(schema_span, message)
    }
}

impl From<graphql_parser::schema::ParseError> for SchemaLoadError {
    fn from(e: graphql_parser::schema::ParseError) -> SchemaLoadError {
        SchemaLoadError::ParseError(e.to_string())
    }
}

impl From<std::io::Error> for SchemaLoadError {
    fn from(e: std::io::Error) -> SchemaLoadError {
        SchemaLoadError::IoError(e.to_string())
    }
}

/// Extension trait for the schema Field type
pub trait FieldExt {
    fn required_arguments(&self) -> Vec<InputValue>;
    fn optional_arguments(&self) -> Vec<InputValue>;
}

impl FieldExt for Field {
    fn required_arguments(&self) -> Vec<InputValue> {
        self.arguments
            .iter()
            .filter(|arg| {
                // Note: We're passing an empty TypeIndex in here, but that's OK as
                // we only want to know if things are required
                FieldArgument::from_input_value(arg, &TypeIndex::empty()).is_required()
            })
            .cloned()
            .collect()
    }

    fn optional_arguments(&self) -> Vec<InputValue> {
        self.arguments
            .iter()
            .filter(|arg| {
                // Note: We're passing an empty TypeIndex in here, but that's OK as
                // we only want to know if things are required
                !FieldArgument::from_input_value(arg, &TypeIndex::empty()).is_required()
            })
            .cloned()
            .collect()
    }
}

/// Extension trait for the schema Type type
pub trait TypeExt {
    fn to_graphql_string(&self) -> String;
    fn inner_name(&self) -> &str;
    fn is_required(&self) -> bool;
}

impl TypeExt for Type {
    fn to_graphql_string(&self) -> String {
        match self {
            Type::NamedType(name) => name.clone(),
            Type::ListType(inner) => format!("[{}]", inner.to_graphql_string()),
            Type::NonNullType(inner) => format!("{}!", inner.to_graphql_string()),
        }
    }

    fn inner_name(&self) -> &str {
        match self {
            Type::NamedType(s) => s,
            Type::ListType(inner) => inner.inner_name(),
            Type::NonNullType(inner) => inner.inner_name(),
        }
    }

    fn is_required(&self) -> bool {
        matches!(self, Type::NonNullType(_))
    }
}

pub trait ScalarTypeExt {
    fn is_builtin(&self) -> bool;
}

impl ScalarTypeExt for ScalarType {
    fn is_builtin(&self) -> bool {
        matches!(self.name.as_ref(), "String" | "Int" | "Boolean" | "ID")
    }
}
