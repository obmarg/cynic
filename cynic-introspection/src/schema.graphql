schema {
  query: Query
}

type Query {
  __schema: __Schema!
  __type(name: String!): __Type
}

type __Schema {
  description: String
  types: [__Type!]!
  queryType: __Type!
  mutationType: __Type
  subscriptionType: __Type
  directives: [__Directive!]!
}

type __Type {
  kind: __TypeKind!
  name: String
  description: String

  # should be non-null for OBJECT and INTERFACE only, must be null for the others
  fields(includeDeprecated: Boolean = false): [__Field!]

  # should be non-null for OBJECT and INTERFACE only, must be null for the others
  interfaces: [__Type!]

  # should be non-null for INTERFACE and UNION only, always null for the others
  possibleTypes: [__Type!]

  # should be non-null for ENUM only, must be null for the others
  enumValues(includeDeprecated: Boolean = false): [__EnumValue!]

  # should be non-null for INPUT_OBJECT only, must be null for the others
  inputFields: [__InputValue!]

  # should be non-null for NON_NULL and LIST only, must be null for the others
  ofType: __Type
}

type __Field {
  name: String!
  description: String
  args: [__InputValue!]!
  type: __Type!
  isDeprecated: Boolean!
  deprecationReason: String
}

type __InputValue {
  name: String!
  description: String
  type: __Type!
  defaultValue: String
}

type __EnumValue {
  name: String!
  description: String
  isDeprecated: Boolean!
  deprecationReason: String
}

enum __TypeKind {
  SCALAR
  OBJECT
  INTERFACE
  UNION
  ENUM
  INPUT_OBJECT
  LIST
  NON_NULL
}

type __Directive {
  name: String!
  description: String
  locations: [__DirectiveLocation!]!
  args: [__InputValue!]!
  isRepeatable: Boolean!
}

enum __DirectiveLocation {
  QUERY
  MUTATION
  SUBSCRIPTION
  FIELD
  FRAGMENT_DEFINITION
  FRAGMENT_SPREAD
  INLINE_FRAGMENT
  VARIABLE_DEFINITION
  SCHEMA
  SCALAR
  OBJECT
  FIELD_DEFINITION
  ARGUMENT_DEFINITION
  INTERFACE
  UNION
  ENUM
  ENUM_VALUE
  INPUT_OBJECT
  INPUT_FIELD_DEFINITION
}
