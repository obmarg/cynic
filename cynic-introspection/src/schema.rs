pub use crate::query::DirectiveLocation;

#[derive(Debug, Clone, PartialEq, Eq)]
#[non_exhaustive]
pub struct Schema {
    pub query_type: String,
    pub mutation_type: Option<String>,
    pub subscription_type: Option<String>,
    pub types: Vec<Type>,
    pub directives: Vec<Directive>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[non_exhaustive]
pub struct Directive {
    pub name: String,
    pub description: Option<String>,
    pub args: Vec<InputValue>,
    pub locations: Vec<DirectiveLocation>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[non_exhaustive]
pub enum Type {
    Object(ObjectType),
    InputObject(InputObjectType),
    Enum(EnumType),
    Interface(InterfaceType),
    Union(UnionType),
    Scalar(ScalarType),
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ObjectType {
    pub name: String,
    pub description: Option<String>,
    pub fields: Vec<Field>,
    pub interfaces: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[non_exhaustive]
pub struct InputObjectType {
    pub name: String,
    pub description: Option<String>,
    pub fields: Vec<InputValue>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[non_exhaustive]
pub struct EnumType {
    pub name: String,
    pub description: Option<String>,
    pub values: Vec<EnumValue>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[non_exhaustive]
pub struct InterfaceType {
    pub name: String,
    pub description: Option<String>,
    pub fields: Vec<Field>,
    pub interfaces: Vec<String>,
    pub possible_types: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[non_exhaustive]
pub struct UnionType {
    pub name: String,
    pub description: Option<String>,
    pub possible_types: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[non_exhaustive]
pub struct ScalarType {
    pub name: String,
    pub description: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[non_exhaustive]
pub struct EnumValue {
    pub name: String,
    pub description: Option<String>,
    pub deprecated: Deprecated,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[non_exhaustive]
pub enum Deprecated {
    No,
    Yes(Option<String>),
}

impl Deprecated {
    fn new(is_deprecated: bool, deprecation_reason: Option<String>) -> Deprecated {
        match (is_deprecated, deprecation_reason) {
            (false, _) => Deprecated::No,
            (true, reason) => Deprecated::Yes(reason),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[non_exhaustive]
pub struct Field {
    pub name: String,
    pub description: Option<String>,
    pub args: Vec<InputValue>,
    pub ty: FieldType,
    pub deprecated: Deprecated,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[non_exhaustive]

pub struct InputValue {
    pub name: String,
    pub description: Option<String>,
    pub ty: FieldType,
    pub default_value: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[non_exhaustive]
pub struct FieldType {
    pub wrapping: FieldWrapping,
    pub name: String,
}

#[derive(Clone, Copy, PartialEq, Eq)]
pub struct FieldWrapping([u8; 8]);

impl std::fmt::Debug for FieldWrapping {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{:?}", self.into_iter().collect::<Vec<_>>())
    }
}

impl IntoIterator for FieldWrapping {
    type Item = WrappingType;

    type IntoIter = Box<dyn Iterator<Item = WrappingType>>;

    fn into_iter(self) -> Self::IntoIter {
        Box::new(
            self.0
                .into_iter()
                .take_while(|&item| item != 0)
                .flat_map(|item| match item {
                    1 => std::iter::once(WrappingType::List),
                    2 => std::iter::once(WrappingType::NonNull),
                    _ => unreachable!(),
                }),
        )
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum WrappingType {
    List,
    NonNull,
}

#[derive(thiserror::Error, Debug, Eq, PartialEq)]
pub enum SchemaError {
    #[error("A type in the introspection output was missing a name")]
    TypeMissingName,
    #[error("Found a list wrapper type in a position that should contain a named type")]
    UnexpectedList,
    #[error("Found a non-null wrapper type in a position that should contain a named type")]
    UnexpectedNonNull,
    #[error("Found a wrapping type that had no inner ofType")]
    WrappingTypeWithNoInner,
    #[error("Found a wrapping type that was too nested")]
    TooMuchWrapping,
}

impl TryFrom<crate::query::Schema> for Schema {
    type Error = SchemaError;

    fn try_from(schema: crate::query::Schema) -> Result<Self, Self::Error> {
        Ok(Schema {
            query_type: schema.query_type.into_name()?,
            mutation_type: schema.mutation_type.map(|ty| ty.into_name()).transpose()?,
            subscription_type: schema
                .subscription_type
                .map(|ty| ty.into_name())
                .transpose()?,
            types: schema
                .types
                .into_iter()
                .map(TryInto::try_into)
                .collect::<Result<Vec<_>, _>>()?,
            directives: schema
                .directives
                .into_iter()
                .map(TryInto::try_into)
                .collect::<Result<Vec<_>, _>>()?,
        })
    }
}

impl TryFrom<crate::query::Type> for Type {
    type Error = SchemaError;

    fn try_from(ty: crate::query::Type) -> Result<Self, Self::Error> {
        match ty.kind {
            crate::query::TypeKind::Scalar => Ok(Type::Scalar(ScalarType {
                name: ty.name.ok_or(SchemaError::TypeMissingName)?,
                description: ty.description,
            })),
            crate::query::TypeKind::Object => Ok(Type::Object(ObjectType {
                name: ty.name.ok_or(SchemaError::TypeMissingName)?,
                fields: ty
                    .fields
                    .unwrap_or_default()
                    .into_iter()
                    .map(TryInto::try_into)
                    .collect::<Result<Vec<_>, _>>()?,
                description: ty.description,
                interfaces: ty
                    .interfaces
                    .unwrap_or_default()
                    .into_iter()
                    .map(|ty| ty.into_name())
                    .collect::<Result<Vec<_>, _>>()?,
            })),
            crate::query::TypeKind::Interface => Ok(Type::Interface(InterfaceType {
                name: ty.name.ok_or(SchemaError::TypeMissingName)?,
                description: ty.description,
                fields: ty
                    .fields
                    .unwrap_or_default()
                    .into_iter()
                    .map(TryInto::try_into)
                    .collect::<Result<Vec<_>, _>>()?,
                interfaces: ty
                    .interfaces
                    .unwrap_or_default()
                    .into_iter()
                    .map(|ty| ty.into_name())
                    .collect::<Result<Vec<_>, _>>()?,
                possible_types: ty
                    .possible_types
                    .unwrap_or_default()
                    .into_iter()
                    .map(|ty| ty.into_name())
                    .collect::<Result<Vec<_>, _>>()?,
            })),
            crate::query::TypeKind::Union => Ok(Type::Union(UnionType {
                name: ty.name.ok_or(SchemaError::TypeMissingName)?,
                description: ty.description,
                possible_types: ty
                    .possible_types
                    .unwrap_or_default()
                    .into_iter()
                    .map(|ty| ty.into_name())
                    .collect::<Result<Vec<_>, _>>()?,
            })),
            crate::query::TypeKind::Enum => Ok(Type::Enum(EnumType {
                name: ty.name.ok_or(SchemaError::TypeMissingName)?,
                description: ty.description,
                values: ty
                    .enum_values
                    .unwrap_or_default()
                    .into_iter()
                    .map(Into::into)
                    .collect(),
            })),
            crate::query::TypeKind::InputObject => Ok(Type::InputObject(InputObjectType {
                name: ty.name.ok_or(SchemaError::TypeMissingName)?,
                description: ty.description,
                fields: ty
                    .input_fields
                    .unwrap_or_default()
                    .into_iter()
                    .map(InputValue::try_from)
                    .collect::<Result<Vec<_>, _>>()?,
            })),
            crate::query::TypeKind::List => Err(SchemaError::UnexpectedList),
            crate::query::TypeKind::NonNull => Err(SchemaError::UnexpectedNonNull),
        }
    }
}

impl TryFrom<crate::query::Field> for Field {
    type Error = SchemaError;

    fn try_from(field: crate::query::Field) -> Result<Self, Self::Error> {
        Ok(Field {
            name: field.name,
            description: field.description,
            args: field
                .args
                .into_iter()
                .map(TryInto::try_into)
                .collect::<Result<Vec<_>, _>>()?,
            ty: field.ty.try_into()?,
            deprecated: Deprecated::new(field.is_deprecated, field.deprecation_reason),
        })
    }
}

impl TryFrom<crate::query::FieldType> for FieldType {
    type Error = SchemaError;

    fn try_from(field_type: crate::query::FieldType) -> Result<Self, Self::Error> {
        let mut wrapping = [0; 8];
        let mut wrapping_pos = 0;
        let mut current_ty = field_type;
        loop {
            if wrapping_pos >= wrapping.len() {
                return Err(SchemaError::TooMuchWrapping);
            }
            match current_ty.kind {
                crate::query::TypeKind::List => {
                    wrapping[wrapping_pos] = 1;
                    wrapping_pos += 1;
                    current_ty = *current_ty
                        .of_type
                        .ok_or(SchemaError::WrappingTypeWithNoInner)?;
                }
                crate::query::TypeKind::NonNull => {
                    wrapping[wrapping_pos] = 2;
                    wrapping_pos += 1;
                    current_ty = *current_ty
                        .of_type
                        .ok_or(SchemaError::WrappingTypeWithNoInner)?;
                }
                _ => {
                    return Ok(FieldType {
                        name: current_ty.name.ok_or(SchemaError::TypeMissingName)?,
                        wrapping: FieldWrapping(wrapping),
                    })
                }
            };
        }
    }
}

impl TryFrom<crate::query::InputValue> for InputValue {
    type Error = SchemaError;

    fn try_from(value: crate::query::InputValue) -> Result<Self, Self::Error> {
        Ok(InputValue {
            name: value.name,
            description: value.description,
            ty: value.ty.try_into()?,
            default_value: value.default_value,
        })
    }
}

impl From<crate::query::EnumValue> for EnumValue {
    fn from(value: crate::query::EnumValue) -> Self {
        EnumValue {
            name: value.name,
            description: value.description,
            deprecated: Deprecated::new(value.is_deprecated, value.deprecation_reason),
        }
    }
}

impl TryFrom<crate::query::Directive> for Directive {
    type Error = SchemaError;

    fn try_from(value: crate::query::Directive) -> Result<Self, Self::Error> {
        Ok(Directive {
            name: value.name,
            description: value.description,
            args: value
                .args
                .into_iter()
                .map(TryInto::try_into)
                .collect::<Result<Vec<_>, _>>()?,
            locations: value.locations,
        })
    }
}

impl crate::query::NamedType {
    fn into_name(self) -> Result<String, SchemaError> {
        self.name.ok_or(SchemaError::TypeMissingName)
    }
}
