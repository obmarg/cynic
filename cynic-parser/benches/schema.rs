// use divan::AllocProfiler;

// #[global_allocator]
// static ALLOC: AllocProfiler = AllocProfiler::system();

fn main() {
    // Run registered benchmarks.
    divan::main();
}

const GITHUB_SCHEMA: &str = include_str!("../../schemas/github.graphql");

#[divan::bench]
fn cynic_parser() -> cynic_parser::TypeSystemDocument {
    let parsed = cynic_parser::parse_type_system_document(GITHUB_SCHEMA).unwrap();
    divan::black_box(parsed)
}

#[divan::bench(
    types = [
        &str,
        String
    ]
)]
fn graphql_parser<T>() -> graphql_parser::schema::Document<'static, T>
where
    T: graphql_parser::query::Text<'static>,
{
    let parsed = graphql_parser::parse_schema(GITHUB_SCHEMA).unwrap();
    divan::black_box(parsed)
}

#[divan::bench]
fn async_graphql_parser() -> async_graphql_parser::types::ServiceDocument {
    let parsed = async_graphql_parser::parse_schema(GITHUB_SCHEMA).unwrap();
    divan::black_box(parsed)
}

#[divan::bench]
fn apollo_parser() -> apollo_parser::SyntaxTree {
    let parser = apollo_parser::Parser::new(GITHUB_SCHEMA);
    let cst = parser.parse();
    divan::black_box(cst)
}
