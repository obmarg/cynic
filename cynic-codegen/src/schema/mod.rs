mod keywords;
mod names;
mod type_index;

pub mod parser;
pub mod types;

mod input;
pub mod markers;

use std::convert::Infallible;
use std::marker::PhantomData;

pub use self::{
    input::SchemaInput,
    names::FieldName,
    parser::{load_schema, Document},
};
use self::{type_index::TypeIndex, types::SchemaRoots};

// TODO: Uncomment this
// pub use self::{types::*},

pub struct Unvalidated;
pub struct Validated;

pub struct Schema<'a, State> {
    type_index: Box<dyn type_index::TypeIndex + 'a>,
    phantom: PhantomData<State>,
}

impl<'a> Schema<'a, Unvalidated> {
    pub fn new(input: SchemaInput) -> Self {
        let type_index: Box<dyn TypeIndex> = match input {
            SchemaInput::Document(document) => {
                Box::new(type_index::SchemaBackedTypeIndex::for_schema(document))
            }
            #[cfg(feature = "rkyv")]
            SchemaInput::Archive(data) => Box::new(
                type_index::optimised::ArchiveBacked::from_checked_data(data),
            ),
        };

        Schema {
            phantom: PhantomData,
            type_index,
        }
    }
}

impl<'a> Schema<'a, Unvalidated> {
    pub fn validate(self) -> Result<Schema<'a, Validated>, SchemaError> {
        self.type_index.validate_all()?;

        Ok(Schema {
            // doc: self.doc,
            type_index: self.type_index,
            phantom: PhantomData,
        })
    }

    pub fn lookup<'b, Kind>(&'b self, name: &str) -> Result<Kind, SchemaError>
    where
        Kind: TryFrom<types::Type<'b>> + 'b,
        Kind::Error: Into<SchemaError>,
    {
        Kind::try_from(self.type_index.lookup_valid_type(name)?).map_err(Into::into)
        // TODO: Suggestion logic should probably be implemented here (or in type_index)
    }
}

impl<'a> Schema<'a, Validated> {
    pub fn root_types(&self) -> Result<SchemaRoots<'_>, SchemaError> {
        self.type_index.root_types()
    }

    pub fn iter(&self) -> impl Iterator<Item = types::Type<'_>> {
        // unsafe_iter is safe because we're in a validated schema
        self.type_index.unsafe_iter()
    }

    // Looks up a kind that we're not certain is in the validated schema.
    pub fn try_lookup<'b, Kind>(&'b self, name: &str) -> Result<Kind, SchemaError>
    where
        Kind: TryFrom<types::Type<'b>, Error = SchemaError>,
    {
        Kind::try_from(self.type_index.lookup_valid_type(name)?)
        // TODO: Suggestion logic should probably be implemented here (or in type_index)
    }

    pub fn lookup<'b, Kind>(&'b self, name: &str) -> Result<Kind, SchemaError>
    where
        Kind: TryFrom<types::Type<'b>>,
        Kind::Error: Into<SchemaError>,
    {
        // unsafe_lookup is safe because we're validated
        Kind::try_from(self.type_index.unsafe_lookup(name).ok_or_else(|| {
            SchemaError::CouldNotFindType {
                name: name.to_string(),
            }
        })?)
        .map_err(Into::into)
    }
}

#[derive(Debug)]
pub enum SchemaError {
    UnexpectedKind {
        name: String,
        expected: types::Kind,
        found: types::Kind,
    },
    InvalidTypeInSchema {
        name: String,
        details: String,
    },
    CouldNotFindType {
        name: String,
    },
}

impl std::fmt::Display for SchemaError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::UnexpectedKind {
                name,
                expected,
                found,
            } => write!(
                f,
                "The type `{}` is a {} in the GraphQL schema but is being used where we expect a {}",
                name, found, expected,
            ),
            Self::InvalidTypeInSchema { name, details } => {
                write!(f, "Invalid schema when validating `{}`: {}", name, details)
            }
            Self::CouldNotFindType { name } => {
                write!(f, "Could not find a type named `{}` in the schema", name)
            }
        }
    }
}

impl From<SchemaError> for crate::Errors {
    fn from(val: SchemaError) -> Self {
        syn::Error::new(proc_macro2::Span::call_site(), val).into()
    }
}

impl SchemaError {
    pub fn unexpected_kind(actual: types::Type<'_>, expected: types::Kind) -> Self {
        SchemaError::UnexpectedKind {
            name: actual.name().to_string(),
            expected,
            found: actual.kind(),
        }
    }
}

impl From<Infallible> for SchemaError {
    fn from(_: Infallible) -> Self {
        panic!("This can't happen")
    }
}
