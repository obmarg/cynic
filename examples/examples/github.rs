//! An example using the GitHub API
//!
//! Note that a lot of this example is feature flagged: this is because rust-analyzer
//! wants to build it as part of the cynic package when I'm working on cynic.  It
//! builds quite slow because of the size of the GitHub API (the schema output is
//! around 100k lines of rust), and it's too much for normal development.
//!
//! If you want to use this example be sure to remove all the feature flagging.
//!
//! This example also requires the `reqwest-blocking` feature to be active.

fn main() {
    #[cfg(feature = "github")]
    {
        let result = run_query();
        println!("{:?}", result);
    }
}

#[cfg(feature = "github")]
fn run_query() -> cynic::GraphQlResponse<queries::PullRequestTitles> {
    use cynic::http::ReqwestBlockingExt;

    let query = build_query();

    let token = std::env::var("GITHUB_TOKEN").expect("GITHUB_TOKEN env var must be set");

    reqwest::blocking::Client::new()
        .post("https://api.github.com/graphql")
        .header("Authorization", format!("Bearer {}", token))
        .header("User-Agent", "obmarg")
        .run_graphql(query)
        .unwrap()
}

#[cfg(feature = "github")]
fn build_query() -> cynic::Operation<queries::PullRequestTitles, queries::PullRequestTitlesArguments>
{
    use cynic::QueryBuilder;
    use queries::{
        IssueOrder, IssueOrderField, OrderDirection, PullRequestTitles, PullRequestTitlesArguments,
    };

    PullRequestTitles::build(PullRequestTitlesArguments {
        pr_order: IssueOrder {
            direction: OrderDirection::Asc,
            field: IssueOrderField::CreatedAt,
        },
    })
}

#[cfg(feature = "github")]
#[cynic::schema_for_derives(file = "../schemas/github.graphql", module = "schema")]
mod queries {
    use super::schema;

    pub type DateTime = chrono::DateTime<chrono::Utc>;

    cynic::impl_scalar!(DateTime, schema::DateTime);

    #[derive(cynic::FragmentArguments, Debug)]
    pub struct PullRequestTitlesArguments {
        pub pr_order: IssueOrder,
    }

    #[derive(cynic::InputObject, Clone, Debug)]
    #[cynic(rename_all = "camelCase")]
    pub struct IssueOrder {
        pub direction: OrderDirection,
        pub field: IssueOrderField,
    }

    #[derive(cynic::Enum, Clone, Copy, Debug)]
    #[cynic(rename_all = "SCREAMING_SNAKE_CASE")]
    pub enum OrderDirection {
        Asc,
        Desc,
    }

    #[derive(cynic::Enum, Clone, Copy, Debug)]
    #[cynic(rename_all = "SCREAMING_SNAKE_CASE")]
    pub enum IssueOrderField {
        Comments,
        CreatedAt,
        UpdatedAt,
    }

    #[derive(cynic::QueryFragment, Debug)]
    #[cynic(graphql_type = "Query", argument_struct = "PullRequestTitlesArguments")]
    pub struct PullRequestTitles {
        #[arguments(name = "cynic".to_string(), owner = "obmarg".to_string())]
        pub repository: Option<Repository>,
    }

    #[derive(cynic::QueryFragment, Debug)]
    #[cynic(argument_struct = "PullRequestTitlesArguments")]
    pub struct Repository {
        #[arguments(orderBy: $pr_order, first: 10)]
        pub pull_requests: PullRequestConnection,
    }

    #[derive(cynic::QueryFragment, Debug)]
    pub struct PullRequestConnection {
        #[cynic(flatten)]
        pub nodes: Vec<PullRequest>,
    }

    #[derive(cynic::QueryFragment, Debug)]
    pub struct PullRequest {
        pub title: String,
        pub created_at: DateTime,
    }
}

#[cfg(feature = "github")]
mod schema {
    cynic::use_schema!("../schemas/github.graphql");
}

#[cfg(all(test, feature = "github"))]
mod test {
    use super::*;

    #[test]
    fn snapshot_test_query() {
        // Running a snapshot test of the query building functionality as that gives us
        // a place to copy and paste the actual GQL we're using for running elsewhere,
        // and also helps ensure we don't change queries by mistake

        let query = build_query();

        insta::assert_snapshot!(query.query);
    }

    #[test]
    fn test_running_query() {
        let result = run_query();
        if result.errors.is_some() {
            assert_eq!(result.errors.unwrap().len(), 0);
        }
        assert_eq!(
            result
                .data
                .as_ref()
                .unwrap()
                .repository
                .as_ref()
                .unwrap()
                .pull_requests
                .nodes
                .len(),
            10
        );
        insta::assert_debug_snapshot!(result.data);
    }
}
