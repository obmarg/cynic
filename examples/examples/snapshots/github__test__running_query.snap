---
source: examples/examples/github.rs
expression: result.data
---
Some(
    PullRequestTitles {
        repository: Some(
            Repository {
                pull_requests: PullRequestConnection {
                    nodes: [
                        PullRequest {
                            title: "Add CI using Github Actions",
                        },
                        PullRequest {
                            title: "Simple type validation in QueryFragment derive.",
                        },
                        PullRequest {
                            title: "Add support for union types via InlineFragments trait.",
                        },
                        PullRequest {
                            title: "Add scalars_as_strings.",
                        },
                        PullRequest {
                            title: "Convert parsing of derive input to use darling",
                        },
                        PullRequest {
                            title: "Fix cynic_arguments parsing with darling",
                        },
                        PullRequest {
                            title: "Added `flatten` option at the field level.",
                        },
                        PullRequest {
                            title: "Optimisations",
                        },
                        PullRequest {
                            title: "First attempt at querygen & web interface",
                        },
                        PullRequest {
                            title: "Added query_module attribute macro.",
                        },
                    ],
                },
            },
        ),
    },
)
